src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1684,22 +1761,117 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                if (this.IsConditionalState)
                 {
-                    // PROTOTYPE(NullReferenceTypes): `ref` arguments should be treated as l-values
-                    // for assignment. See `ref x3` in StaticNullChecking.PassingParameters_01.
-                    VisitRvalue(argument);
+                    // We could be in a conditional state because of NotNullWhenFalse annotation
+                    // Then WhenTrue/False states correspond to the invocation returning true/false
+
+                    // We'll assume that we're in the unconditional state where the method returns true,
+                    // then we'll repeat assuming the method returns false.
+
+                    LocalState whenTrue = this.StateWhenTrue.Clone();
+                    LocalState whenFalse = this.StateWhenFalse.Clone();
+
+                    this.SetState(whenTrue);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                    Debug.Assert(!IsConditionalState);
+                    whenTrue = this.State; // LocalState may be a struct
+
+                    this.SetState(whenFalse);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                    Debug.Assert(!IsConditionalState);
+                    whenFalse = this.State; // LocalState may be a struct
+
+                    this.SetConditionalState(whenTrue, whenFalse);
                 }
                 else
                 {
-                    VisitLvalue(argument);
+                    VisitArgumentEvaluate(arguments, refKindsOpt, i);
+                }
+
+                var argument = arguments[i];
+                if (argument.Type?.IsReferenceType != true)
+                {
+                    continue;
+                }
+
+                int slot = MakeSlot(argument);
+                if (slot <= 0)
+                {
+                    continue;
+                }
+
+                AttributeAnnotations annotation = annotations[i];
+                bool notNullWhenFalse = (annotation & AttributeAnnotations.NotNullWhenFalse) != 0;
+                bool ensuresNotNull = (annotation & AttributeAnnotations.EnsuresNotNull) != 0;
+
+                if (ensuresNotNull)
+                {
+                    // The variable in this slot is not null
+                    if (this.IsConditionalState)
+                    {
+                        this.StateWhenTrue[slot] = true;
+                        this.StateWhenFalse[slot] = true;
+                    }
+                    else
+                    {
+                        this.State[slot] = true;
+                    }
+                }
+                else if (notNullWhenFalse)
+                {
+                    // We'll use the WhenTrue/False states to represent whether the invocation returns true/false
+                    Split();
```

Comment:
>Split(); [](start = 20, length = 8)

I think think this might be a wrong place to do the Split, it should be done when we are visiting invocation, not the arguments.

Reply:
I don't see a functional advantage to splitting early. Do you mean that would be better for clarity?